In [9]:
from sqlalchemy import create_engine
import pandas as pd

In [10]:
import psycopg2

conn = psycopg2.connect(
    host="localhost",
    database="postgres",
    user="postgres",
    password="Arya@980"
)

print("Connected successfully!")

Connected successfully!


In [11]:
engine = create_engine('postgresql+psycopg2://postgres:Arya%40980@localhost:5432/postgres')
print("Connected to SQL!")

Connected to SQL!


In [12]:
query = """
SELECT table_name 
FROM information_schema.tables 
WHERE table_schema = 'public';
"""
pd.read_sql(query, engine)

,table_name
0,ev_data


In [13]:
from sqlalchemy import create_engine
import pandas as pd

engine = create_engine('postgresql+psycopg2://postgres:Arya%40980@localhost:5432/postgres')

print("Connected")

Connected


In [14]:
import pandas as pd

df = pd.read_sql("SELECT * FROM ev_data", engine)

df.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,DOL Vehicle ID,Electric Utility,2020 Census Tract,Vehicle Age,EV Category,Range Group
0,JN1AZ0CP5C,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,BEV,Eligible,73.0,153331706,AVISTA CORP,53065950200.0,14,BEV,Low
1,JTMABABA7P,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,BEV,Unknown,0.0,253586308,PACIFICORP,53077001100.0,3,BEV,Low
2,1N4AZ1CP1J,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,BEV,Eligible,151.0,333135022,CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),53033007800.0,8,BEV,Medium
3,5UX43EU09S,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,PHEV,Eligible,40.0,267525737,PUGET SOUND ENERGY INC,53035940100.0,1,PHEV,Low
4,3C3CFFGE5F,Thurston,Yelm,WA,98597.0,2015,FIAT,500,BEV,Eligible,87.0,474468501,PUGET SOUND ENERGY INC,53067012422.0,11,BEV,Low


In [15]:
df.rename(columns={
    'EV Category': 'ev_category',
    'Vehicle Age': 'vehicle_age',
    'Range Group': 'range_group',
    'Electric Vehicle Type': 'ev_type',
    'Electric Range': 'electric_range',
    'Clean Alternative Fuel Vehicle (CAFV) Eligibility': 'cafv_eligibility'
}, inplace=True)

df.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,ev_type,cafv_eligibility,electric_range,DOL Vehicle ID,Electric Utility,2020 Census Tract,vehicle_age,ev_category,range_group
0,JN1AZ0CP5C,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,BEV,Eligible,73.0,153331706,AVISTA CORP,53065950200.0,14,BEV,Low
1,JTMABABA7P,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,BEV,Unknown,0.0,253586308,PACIFICORP,53077001100.0,3,BEV,Low
2,1N4AZ1CP1J,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,BEV,Eligible,151.0,333135022,CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),53033007800.0,8,BEV,Medium
3,5UX43EU09S,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,PHEV,Eligible,40.0,267525737,PUGET SOUND ENERGY INC,53035940100.0,1,PHEV,Low
4,3C3CFFGE5F,Thurston,Yelm,WA,98597.0,2015,FIAT,500,BEV,Eligible,87.0,474468501,PUGET SOUND ENERGY INC,53067012422.0,11,BEV,Low


In [16]:
query = """
SELECT COUNT(*) AS total_vehicles
FROM ev_data;
"""
pd.read_sql(query, engine)

,total_vehicles
0,279780


# Insight:
The dataset contains a total of 279,780 electric vehicles, providing a large and reliable sample size for analyzing market trends, consumer preferences, and EV adoption patterns.

# EV Type Distribution

In [17]:
query = """
SELECT ev_category, COUNT(*) AS total
FROM ev_data
GROUP BY ev_category;
"""
pd.read_sql(query, engine)

ProgrammingError: (psycopg2.errors.UndefinedColumn) column "ev_category" does not exist
LINE 2: SELECT ev_category, COUNT(*) AS total
               ^

[SQL: 
SELECT ev_category, COUNT(*) AS total
FROM ev_data
GROUP BY ev_category;
]
(Background on this error at: https://sqlalche.me/e/20/f405)

# Insight:
Battery Electric Vehicles (BEVs) dominate the market with around 80% share, significantly higher than Plug-in Hybrid Electric Vehicles (PHEVs). This indicates a strong industry shift toward fully electric vehicles, driven by advancements in battery technology, environmental concerns, and supportive government policies.

# Average Electric Range

In [ ]:
query = """
SELECT ROUND(AVG(electric_range)::numeric, 2) AS avg_range
FROM ev_data;
"""
pd.read_sql(query, engine)

# Insight:
The average electric range of EVs is approximately 39 miles, indicating that a significant portion of vehicles in the dataset are either older models or plug-in hybrid vehicles (PHEVs) with lower battery capacity. This suggests that while EV adoption is growing, there is still a mix of low-range and high-range vehicles in the market, reflecting the transition phase of EV technology.

# Top 10 Companies

In [ ]:
query = """
SELECT "Make", COUNT(*) AS total
FROM ev_data
GROUP BY "Make"
ORDER BY total DESC
LIMIT 10;
"""
pd.read_sql(query, engine)

# Insight:
Tesla is the clear market leader with a significantly higher number of vehicles compared to all other manufacturers. The large gap between Tesla and the second-largest player (Chevrolet) indicates a highly concentrated market where Tesla holds a dominant position. This reflects Tesla’s strong brand value, early mover advantage, and leadership in EV technology, while traditional automakers are still catching up.

# Market Share %

In [ ]:
query = """
SELECT 
    "Make",
    COUNT(*) AS total,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2) AS market_share
FROM ev_data
GROUP BY "Make"
ORDER BY total DESC
LIMIT 10;
"""
pd.read_sql(query, engine)

# Insight:
Tesla holds a dominant market share of over 41%, which is significantly higher than any other competitor. The sharp drop to the second-largest player (~7%) highlights a highly concentrated EV market with limited competition at the top. This suggests that Tesla has a strong competitive advantage in terms of technology, infrastructure, and brand positioning, making it difficult for other manufacturers to capture significant market share.

# Year-over-Year Growth (YOY)

In [ ]:
query = """
SELECT 
    "Model Year",
    COUNT(*) AS total,
    LAG(COUNT(*)) OVER (ORDER BY "Model Year") AS prev_year,
    ROUND(
        (COUNT(*) - LAG(COUNT(*)) OVER (ORDER BY "Model Year")) * 100.0 
        / LAG(COUNT(*)) OVER (ORDER BY "Model Year"), 
    2) AS yoy_growth
FROM ev_data
GROUP BY "Model Year"
ORDER BY "Model Year";
"""
pd.read_sql(query, engine)

# Insight:
While the data shows negative growth in recent years, this is primarily due to incomplete or partially available data for newer model years rather than an actual decline in EV adoption. The overall trend still indicates strong long-term growth, with significant acceleration after 2018 and peak adoption around 2023.

# Company Strategy (BEV vs PHEV)

In [23]:
query = """
SELECT 
    "Make",
    "EV Category",
    COUNT(*) AS total
FROM ev_data
GROUP BY "Make", "EV Category"
ORDER BY "Make", total DESC;
"""

df = pd.read_sql(query, engine)
df.head()

,Make,EV Category,total
0,ACURA,BEV,356
1,ALFA ROMEO,PHEV,104
2,AUDI,BEV,3898
3,AUDI,PHEV,2039
4,AZURE DYNAMICS,BEV,3


# Insight:
Audi demonstrates a balanced electrification strategy with a stronger focus on BEVs, indicating a gradual transition toward full electric adoption. 

In contrast, Acura and Alfa Romeo show limited EV presence, while the overall trend highlights clear industry dominance of BEVs over PHEVs.

# Top Models per Company

In [24]:
query = """
SELECT *
FROM (
    SELECT 
        "Make",
        "Model",
        COUNT(*) AS total,
        ROW_NUMBER() OVER (PARTITION BY "Make" ORDER BY COUNT(*) DESC) AS rank
    FROM ev_data
    GROUP BY "Make", "Model"
) sub
WHERE rank <= 3
ORDER BY "Make", rank;
"""
pd.read_sql(query, engine)

,Make,Model,total,rank
0,ACURA,ZDX,356,1
1,ALFA ROMEO,TONALE,104,1
2,AUDI,Q5 E,1329,1
3,AUDI,E-TRON,1257,2
4,AUDI,Q4,1069,3
...,...,...,...,...
100,VOLKSWAGEN,ID. BUZZ,415,3
101,VOLVO,XC90,2438,1
102,VOLVO,XC60,1973,2
103,VOLVO,XC40,1413,3


# Insight:
The analysis reveals that each manufacturer has a few dominant models driving the majority of their EV sales. For example, companies like Audi and Volvo show a clear concentration in their top 2–3 models, indicating a focused product strategy. This suggests that EV adoption is heavily influenced by flagship models rather than a wide distribution across all offerings, highlighting the importance of successful key models in driving market share.

# Range Segmentation

In [26]:
query = """
SELECT 
    CASE 
        WHEN "Electric Range" = 0 THEN 'No Range (PHEV/Unknown)'
        WHEN "Electric Range" < 100 THEN 'Low Range'
        WHEN "Electric Range" BETWEEN 100 AND 250 THEN 'Medium Range'
        ELSE 'High Range'
    END AS range_segment,
    COUNT(*) AS total
FROM ev_data
GROUP BY 1
ORDER BY total DESC;
"""

df_range = pd.read_sql(query, engine)
df_range

,range_segment,total
0,No Range (PHEV/Unknown),177948
1,Low Range,64237
2,Medium Range,27938
3,High Range,9657


# Insight:
A significant majority of vehicles fall under the "No Range" or low-range categories, indicating that a large portion of the dataset consists of PHEVs or older EV models with limited battery capacity. Only a small percentage of vehicles fall into the high-range category, suggesting that advanced long-range EV technology is still emerging. This reflects a transitional market where adoption is increasing, but high-performance EV penetration remains relatively low.

# Cohort-style Analysis (EV Adoption by Year Type)

In [16]:
query = """
SELECT 
    "Model Year",
    ev_category,
    COUNT(*) AS total,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (PARTITION BY "Model Year"), 2) AS percentage
FROM ev_data
GROUP BY "Model Year", ev_category
ORDER BY "Model Year", ev_category;
"""
pd.read_sql(query, engine)

,Model Year,ev_category,total,percentage
0,1999,BEV,2,100.00
1,2000,BEV,7,100.00
2,2002,BEV,1,100.00
3,2003,BEV,1,100.00
4,2008,BEV,19,100.00
5,2010,BEV,20,90.91
6,2010,PHEV,2,9.09
7,2011,BEV,521,90.14
8,2011,PHEV,57,9.86
9,2012,BEV,607,44.21


# Insight:
The cohort analysis shows a clear transition in the EV market from a balanced mix of BEVs and PHEVs to a strong dominance of BEVs over time. In earlier years (2012–2014), PHEVs held a significant share, even exceeding BEVs in some cases, indicating an initial reliance on hybrid technology. However, from 2018 onwards, BEVs consistently dominate with over 80% share in recent years, reflecting advancements in battery technology, improved charging infrastructure, and growing consumer preference for fully electric vehicles. This highlights a decisive industry shift toward full electrification.